Spack extensions can import code from modules in root or cmd folder

alalazo opened the pull request:
This PR is a follow up of #8612 and an alternative implementation of the improved import capabilities in #11145.

The problem with #8612 was basically that for complex commands one couldn't split source code into python modules that were residing:
1. In the extension root folder 
2. In a subfolder of the `cmd` directory

This PR solves that issue by ensuring that, for each extension found, the python packages associated with the root folder and the `cmd` folder are in `sys.modules` with an appropriate `__path__` set.

List of modifications:
- [x] Spack extensions can import code from modules in root or cmd folder (either relative or absolute imports)
- [x] Added unit tests to verify this behavior

alalazo commented on lib/spack/spack/extensions.py:
```diff
@@ -50,11 +50,37 @@ def load_command_extension(command, path):
+    # Create a searchable package for both the root folder of the extension
+    # and the subfolder containing the commands
+    ensure_package_creation(extension)
+    ensure_package_creation(extension + '.cmd')
```
The two lines above create two python packages named `spack.extensions.<extension-name>` and `spack.extensions.<extension-name>.cmd`. This means that:
- The `__init__.py` file in those two extension folders is always executed when Spack imports commands
- A user can refer to python modules either in the extension root or in any subfolder of the `<ext-root>/cmd` directory

alalazo commented on lib/spack/spack/test/extensions.py:
```diff
@@ -0,0 +1,98 @@
+
+# Test a relative import
+from ..implementation import hello_folks
+
```
The test here makes sure that both absolute and relative imports work.

alalazo commented:
@scheibelp @chissg If you are fine with the changes above I'll add docs for the extensions with this PR

scheibelp commented:
@alalazo I didn't see discussion from https://github.com/spack/spack/pull/11145 which suggested that it was not salvageable, so I'm curious why you added an alternative implementation.

That being said, this does have tests. So @chissg if it seems suitable for your purposes, it is attractive to me for at least that reason.

alalazo commented:
> I didn't see discussion from #11145 which suggested that it was not salvageable, so I'm curious why you added an alternative implementation.

In the review I recognized the lack of unit tests in #8612, and started this PR to add them. Looking at the code linked in https://github.com/spack/spack/pull/11145#discussion_r275460567 it seemed to me that what was missing from #8612  - please @chissg correct me if I'm wrong - was the possibility to import code outside of the `cmd` folder. So I added a unit test that creates on the fly an extension with the following layout:
```console
spack-testcommand
    └── testcommand
        ├── cmd
        │   └── hello.py
        ├── implementation.py
        └── __init__.py
```
where the command in `hello.py` refers both by relative and absolute imports to functions in `implementation.py`. This unit test was failing with the current `develop` branch and the proposed solution in #11145 was changing a few APIs in `spack/cmd/__init__.py` and `spack/extensions.py` + adding two or three global variables at module scope.

At that point I started reading through the mess that is the import mechanism in Python and figured out a less invasive solution to this based on [package path rules](https://docs.python.org/3/reference/import.html#package-path-rules), implemented it to check it worked and committed it here.

Besides the fact that there are no API changes and no additional global variables added, I think the biggest difference this PR brings wrt #11145 is that by using package path rules instead of global `sys.path` like in:

https://github.com/spack/spack/blob/831f829e696369501d18fb208ea27b0fe51d67fa/lib/spack/spack/extensions.py#L60-L75

we effectively insulate extensions from one other, while having all the paths added in `sys.path` an extension can possibly refer to another depending on the import order.

alalazo commented:
@scheibelp To add on the comment above, #11145 improves error handling for missing or mistyped commands - that part is not present in this PR, I just dealt with imports.

alalazo commented:
> What happens if the __init__.py file doesn't exist in either case? Is it created, or treated as if it existed but was empty, or is this an error?

It's treated as if empty, the code that deals with it is here:
https://github.com/spack/spack/blob/9b1f1ca05c84248e144444aae6e7fe680f1a6dae/lib/spack/spack/extensions.py#L58-L63

alalazo commented:
> Would it be possible to extend your test to verify behavior in this case? Specifically import a command, call one command to set a global and another to retrieve it, verifying that the value retrieved is as set?

Sure. Can you be more specific on what you mean by "call one command to set a global and another to retrieve it"? Going back to the example above, do you mean something like:

1. Call from `hello.py` a function that sets a global in `implementation.py`
2. Then call from `hello.py` a function that retrieves it and check the value

?

chissg commented:
In the specific case I have in mind, [dev.py](https://github.com/chissg/spackdev/blob/7dabe432a1fc92b6e55af5457b27feb2def07f2b/dev/cmd/dev.py), globals `_subcmds` and `_subcmd_functions` are set by a call to `setup_parser()` and then used by a subsequent call to `dev()`. In the normal way in which commands are initialized by Spack, I don't actually know whether these calls are made to the same module object, or to two module objects obtained by distinct import commands.

alalazo commented:
@chissg The test case requested should be in d1934f6. One thing that was not clear, at least to me, when reading the [API docs](https://docs.python.org/3/library/importlib.html) is that if you don't save the loaded module in `sys.modules` global objects won't persist.

scheibelp commented on lib/spack/spack/test/extensions.py:
```diff
@@ -0,0 +1,110 @@
+    del sys.modules['spack.extensions.testcommand']
+    del sys.modules['spack.extensions.testcommand.implementation']
+    del sys.modules['spack.extensions.testcommand.cmd']
+    del sys.modules['spack.extensions.testcommand.cmd.hello']
```
@alalazo if I pause execution just after this line, I see some remaining related entries in `sys.modules`: `spack.extensions.testcommand.__builtin__`, `spack.extensions.testcommand._pytest`, and `spack.extensions.testcommand.cmd.spack`. I'm wondering if we could save the keys we have before doing any of the imports, and then delete anything extra at this point. While these modules are likely harmless (the `__builtin__` module may have odd side effects though), I prefer to clean everything because it makes sure we are tracking the side effects.

alalazo replied:
Done

scheibelp commented:
In addition to addressing https://github.com/spack/spack/pull/11209#discussion_r280964643, could you also rename the `extensions` test module to `cmd_extensions` to clarify that it is for testing command extensions rather than Spack extension packages? To be clear, I'm not asking you to rename `lib/spack/spack/extensions.py`, just `lib/spack/spack/test/extensions.py`.

alalazo commented:
> In addition to addressing #11209 (comment), could you also rename the extensions test module to cmd_extensions [...] lib/spack/spack/test/extensions.py.

No prob renaming that, but I'd like later to improve this feature to allow for:
1. Defining hooks in extensions
2. Defining new configuration files in extensions, if need be
3. Maybe have a cli interface to help managing external commands etc.

so I'm wondering whether we should look for a more general name than `cmd_extensions` (or maybe use `package_extensions` if we add test for extension packages?). In any case, it's fine with me if we want to revisit this discussion later and for the time being just rename the test file to `cmd_extensions.py`.

scheibelp commented:
I'd prefer it be named `cmd_extensions` for now.

I think that name would also be suitable later (even if the functionality you mention in https://github.com/spack/spack/pull/11209#issuecomment-489428653 is added).

> maybe use package_extensions if we add test for extension packages?

Now that there are two types of extensions (extension packages, and command extensions), it would be best to be explicit about both. In the future I think the `extensions` module in `lib/spack/spack` should also be renamed for clarity. And if extension packages were tested in their own module in the future (for now I think there are tests but in the `Package`-testing logic) then it should  be called `package_extensions`.

scheibelp commented on lib/spack/spack/test/extensions.py:
```diff
@@ -0,0 +1,111 @@
+    assert 'Hello world!' in output
+
+
+def test_subcommand_in_nested_directory(hello_world_with_module_in_root):
```
I think a clearer name for this would be `test_command_with_root_import` or `test_command_with_import`: I think you are referring to the imported function as the (sub)command. IMO this is confusing because the focus of these tests is to verify the behavior of "command" in the sense of the command extension.

In other words, the fact that you are importing a function isn't the thing this test is actually testing, it's whether you can import anything from the defined module. I don't think the "nested_directory" suffix adds clarity either, so I also removed that from the names I suggested.

alalazo replied:
Renamed file and function.

alalazo commented:
@scheibelp Done!

scheibelp commented on lib/spack/spack/extensions.py:
```diff
@@ -50,14 +50,47 @@ def load_command_extension(command, path):
-    module_name = '{0}.{1}'.format(__name__, python_name)
 
     try:
+        # TODO: Upon removal of support for Python 2.6 substitute the call
```
@alalazo sorry - another question: does this apply to all calls to `load_module_from_file`, or just this one? If it's just this one, then why is that?

alalazo replied:
It's just this one. The reason has to do with how the import mechanism works (I had to read carefully through the docs to understand that). This snippet:

https://github.com/spack/spack/blob/cdfcf7aa76500f8ac519a31d2dc576f0ccda4226/lib/spack/spack/extensions.py#L72-L78

is used to add a search `__path__` to both the "root" extension module and its  command submodule, e.g. `spack.extensions.testcommand` and `spack.extensions.testcommand.cmd`, which are imported here:

https://github.com/spack/spack/blob/cdfcf7aa76500f8ac519a31d2dc576f0ccda4226/lib/spack/spack/extensions.py#L80-L83

After we do this, we won't need the absolute path of each command file anymore, as that will be computed from the `__path__` of the corresponding subpackage.  For instance, if I import `spack.extensions.testcommand.cmd.hello` the import mechanism will compute the corresponding file path from the first `__path__` attribute found going up with subpackages i.e. `spack.extensions.testcommand.cmd`.

Now, in python 2.6 to import a module based on a computed string we need either to use directly the `__import__` function or the `imp` module - and both approaches are not recommended. From python 2.7 on there's `importlib.import_module` which seems to be the idiomatic way and the comment above is a reminder that we could just use that instead of loading the module and adding it to `sys.modules`.

alalazo replied:
@scheibelp Let me know if something should be clarified in the comment above or if it's fine like that. Thanks!

scheibelp commented:
Thanks!
